impl/client/src/main/javascript/web/prompting/PromptPanel.js

```diff
@@ -34,1658 +33,1728 @@
+                component.valuesArray = newValuesArray;
+                updateNeeded = true;
+              }
 
-        /**
-         * Registers a post init event on the dashboard
-         * @param {Function} callback The function to be executed when the event is triggered
-         */
-        onPostInit: function(callback) {
-          this.getDashboard().on('cdf:postInit', callback);
-        },
+              if(this.autoSubmit) {
+                this.forceSubmit = true;
+              }
 
-        /**
-         * Sets the parameter definition for the prompt panel. Also sets whether the prompt panel has auto submit
-         * @param paramDefn {Object} The parameter definition object
-         */
-        setParamDefn: function(paramDefn) {
-          var prevParamDefn = this.paramDefn;
-          this.paramDefn = paramDefn;
+              if(!updateNeeded) {
+                var paramSelectedValues = param.getSelectedValuesValue();
+                var dashboardParameter = this.dashboard.getParameterValue(component.parameter);
 
-          var fireStateChanged = function(paramName, oldParamDefn, newParamDefn, getValueCallback) {
-            if (this.onStateChanged == null) {
-              return;
-            }
+                // if the dashboardParameter is not an array, paramSelectedValues shouldn't be either
+                if(!_.isArray(dashboardParameter) && paramSelectedValues.length == 1) {
+                  paramSelectedValues = paramSelectedValues[0];
+                }
 
-            var oldVal = oldParamDefn ? getValueCallback(oldParamDefn) : undefined;
-            var newVal = newParamDefn ? getValueCallback(newParamDefn) : undefined;
+                updateNeeded = _areParamsDifferent(dashboardParameter, paramSelectedValues, param.type);
+                if(!updateNeeded && param.isErrorChanged) {
+                  updateNeeded = true;
+                }
+              }
 
-            if (oldVal != newVal) {
-              this.onStateChanged(paramName, oldVal, newVal);
+              if(updateNeeded) {
+                var groupPanel = this.dashboard.getComponentByName(groupName);
+                _mapComponents(groupPanel, function(component) {
+                  this.dashboard.updateComponent(component);
+                }.bind(this));
+              }
             }
-          }.bind(this);
+          }
+        }
+      },
 
-          if (paramDefn) {
-            if(this.autoSubmit == undefined) {
-              this.setAutoSubmit(paramDefn.allowAutoSubmit());
-            }
+      /**
+       * Updates the dashboard and prompt panel based off of differences in the parameter definition
+       *
+       * @method update
+       * @param {JSON} diff - contains the differences between the old and new parameter definitions produced by ParameterDefinitionDiffer.diff
+       */
+      update: function(diff) {
+        var toRemove = Object.keys(diff.toRemove).length > 0,
+          toAdd 
```

Comment:
`isRefresh` and `isForceRefresh` sound like Boolean variables. Can't they be initialised to `false`? If not, why `null` and `undefined`?